### PR TITLE
add specializations for some triangular-triangular multiplications

### LINKED ISSUE #1486: CUBLAS: Multiplication of `UpperTriangular`/`LowerTriangular` not supported
**Describe the bug**

Matrix multiplications with `UpperTriangular` and `LowerTriangular` are outrageously slow.

```julia
using CUDA
using BenchmarkTools

julia> @benchmark U*U setup=(U=triu(CUDA.randn(1024, 1024)))
BenchmarkTools.Trial: 581 samples with 6 evaluations.
 Range (min … max):  5.993 μs … 148.436 ms  ┊ GC (min … max): 0.00% … 0.22%
 Time  (median):     6.445 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.026 ms ±  12.242 ms  ┊ GC (mean ± σ):  0.21% ± 0.02%

  █▇▄▆▂                                                        
  █████▄▆▅▄██▇██▁▅▄▆▆▇▆▇▇▁▆▆▆▄▄▅▅▆▁▆▅▄▁▆▆▄▄▁▄▄▁▄▁▁▁▁▁▁▁▁▄▄▁▁▄ ▇
  5.99 μs      Histogram: log(frequency) by time      22.4 μs <

 Memory estimate: 640 bytes, allocs estimate: 32.
```

```julia
julia> @benchmark U*U setup=(U=UpperTriangular(CUDA.randn(1024, 1024)))
BenchmarkTools.Trial: 2259 samples with 1 evaluation.
 Range (min … max):  26.422 μs … 135.895 ms  ┊ GC (min … max): 0.00% … 1.08%
 Time  (median):      1.942 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):    2.104 ms ±   5.660 ms  ┊ GC (mean ± σ):  0.12% ± 0.05%

  ▆                               █▅ ▅ ▂▅     ▁▂ ▂▂         ▃▁ ▁
  █▅▁▁▁▃▁▃▁▁▁▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▃▇██▇████▅▁▃▇▆██▆██▆▆▆▅▆▆▄▆███ █
  26.4 μs       Histogram: log(frequency) by time      3.51 ms <

 Memory estimate: 720 bytes, allocs estimate: 15.
```

```julia
julia> @benchmark U*U' setup=(U=UpperTriangular(CUDA.randn(1024, 1024)))
┌ Warning: Performing scalar indexing on task Task (runnable) @0x00007f5d2af14010.
│ Invocation of getindex resulted in scalar indexing of a GPU array.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on the GPU, but very slowly on the CPU,
│ and therefore are only permitted from the REPL for prototyping purposes.
│ If you did intend to index this array, annotate the caller with @allowscalar.
└ @ GPUArrays ~/.julia/packages/GPUArrays/Zecv7/src/host/indexing.jl:56
BenchmarkTools.Trial: 1 sample with 1 evaluation.
 Single result which took 16.270 s (0.41% GC) to evaluate,
 with a memory estimate of 966.18 MiB, over 6397082 allocations.
```

```julia
julia> @benchmark U'*U setup=(U=UpperTriangular(CUDA.randn(1024, 1024)))
BenchmarkTools.Trial: 2235 samples with 1 evaluation.
 Range (min … max):  27.548 μs … 150.721 ms  ┊ GC (min … max): 0.00% … 1.48%
 Time  (median):      1.925 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):    2.127 ms ±   5.825 ms  ┊ GC (mean ± σ):  0.15% ± 0.05%

  ▆                               ▃█ ▂▄▄▃▄▄▂▁  ▁▂▃▃▁   ▁▂▂     ▁
  █▆▁▄▁▁▁▁▁▁▁▁▃▃▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▃██▆████████▇██████▆▆▃████▆▆▆ █
  27.5 μs       Histogram: log(frequency) by time      3.42 ms <

 Memory estimate: 704 bytes, allocs estimate: 14.
```

```julia
julia> @benchmark U'*U' setup=(U=UpperTriangular(CUDA.randn(1024, 1024)))

BenchmarkTools.Trial: 1 sample with 1 evaluation.
 Single result which took 18.099 s (0.41% GC) to evaluate,
 with a memory estimate of 966.48 MiB, over 6402181 allocations.
```

* `U*U` is slow but works
* `U*L` is outrageous
* `L*U` is slow but works
* `L*L` is outrageous 
In general, using `UpperTriangular` and `LowerTriangular` for matrix multiplications is terribly slow. 
A quick workaround is to use `triu`, `tril` and then put the result back in the adaptors.

```
@v1.7) pkg> status
      Status `~/.julia/environments/v1.7/Project.toml`
  [6e4b80f9] BenchmarkTools v1.3.1
  [052768ef] CUDA v3.9.0
```

</p>
</details>

**Version info**

```
julia> versioninfo()
Julia Version 1.7.2
Commit bf53498635 (2022-02-06 15:21 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 8
```
```
julia> CUDA.versioninfo()
CUDA toolkit 11.6, artifact installation
NVIDIA driver 510.60.2, for CUDA 11.6
CUDA driver 11.6

Libraries: 
- CUBLAS: 11.8.1
- CURAND: 10.2.9
- CUFFT: 10.7.0
- CUSOLVER: 11.3.2
- CUSPARSE: 11.7.1
- CUPTI: 16.0.0
- NVML: 11.0.0+510.60.2
- CUDNN: missing
```

### PR DESCRIPTION
**Pull request by Red-Portal**
This implements a few of the triangular-triangular multiplications by invoking `trmm!`, which addresses #1486

### CODE REVIEW

File: lib/cublas/linalg.jl [hunk 1]
```diff
@@ -443,6 +443,68 @@ for (t, uploc, isunitc) in ((:LowerTriangular, 'U', 'N'),
     end
 end
 
+function LinearAlgebra.mul!(X::DenseCuMatrix{T},
+                            A::LowerTriangular{T,<:DenseCuMatrix},
+                            B::UpperTriangular{T,<:DenseCuMatrix},
+                            ) where {T<:CublasFloat}
+    triu!(parent(B))
+    CUBLAS.trmm!('L', 'L', 'N', 'N', one(T), parent(A), parent(B), parent(X))
```
Comment (maleadt) [inline lib/cublas/linalg.jl]: You shouldn't need CUBLAS module prefixes here, as this is part of the CUBLAS module.

Comment (Red-Portal) [reply lib/cublas/linalg.jl]: Oh, it's only required for the `eval` blocks okay.

File: lib/cublas/linalg.jl [hunk 2]
```diff
@@ -443,6 +443,68 @@ for (t, uploc, isunitc) in ((:LowerTriangular, 'U', 'N'),
...
+            CUBLAS.trmm!('R', 'U', 'T', 'N', one(T), parent(parent(B)), parent(A), parent(X))
+            X
+        end
+
+        function LinearAlgebra.mul!(X::DenseCuMatrix{T},
+                                    A::UpperTriangular{<:Any,<:$trtype{T,<:DenseCuMatrix}},
+                                    B::LowerTriangular{T,<:DenseCuMatrix},
+                                    ) where {T<:$valtype}
+            tril!(parent(B))
+            CUBLAS.trmm!('L', 'L', 'T', 'N', one(T), parent(parent(A)), parent(B), parent(X))
```
Comment (maleadt) [inline lib/cublas/linalg.jl:483]: Same here, and others below.

Comment (Red-Portal) [reply lib/cublas/linalg.jl:483]: @maleadt But the other existing routines use the BLAS prefix explicitly. For example:

```julia
@inline function LinearAlgebra.mul!(y::CuVector{T}, A::Hermitian{T,<:CuMatrix}, x::CuVector{T},
             α::Number, β::Number) where {T<:CublasComplex}
    alpha, beta = promote(α, β, zero(T))
    if alpha isa Union{Bool,T} && beta isa Union{Bool,T}
        return CUBLAS.hemv!(A.uplo, alpha, A.data, x, beta, y)
    else
        error("only supports BLAS type, got $T")
    end
end
```

in line 520. Is there a reason to use a different convention?

Comment (maleadt) [reply lib/cublas/linalg.jl:483]: Ah, I see, those are leftovers from when moving this code. I guess we can merge this as-is and do some clean-up later.

**Comment by Red-Portal** [comment]
Great! Thanks for the quick review.

**Pull request merged**